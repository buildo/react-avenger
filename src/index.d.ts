import { Queries, Commands } from 'avenger';
import { ObjectOmit } from 'typelevel-ts';

type QueriesProps<Decl extends Queries> = { [k in keyof Decl]: Decl[k]['_A'] }[keyof Decl];

type QueriesInnerProps<Decl extends Queries> = {
  [k in keyof Decl]?: Decl[k]['_P']
} & {
  readyState: { [k in keyof Decl]: { loading: boolean, ready: boolean } }
};

type QueriesOuterProps<InnerProps extends {}, Decl extends Queries> = ObjectOmit<InnerProps, keyof Decl | 'readyState'> & QueriesProps<Decl>;

type DeclareQueriesReturn<Decl extends Queries> = (
  <InnerProps extends QueriesInnerProps<Decl>>(Component: React.ComponentType<InnerProps>) => React.ComponentType<QueriesOuterProps<InnerProps, Decl>>
) & { Props: QueriesInnerProps<Decl> };

export function declareQueries<Decl extends Queries>(declaration: Decl): DeclareQueriesReturn<Decl>;
