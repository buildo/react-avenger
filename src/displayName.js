export default wrapper => Component => `${wrapper}(${Component.displayName || Component.name || 'Component'})`;