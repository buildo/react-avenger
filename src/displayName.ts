export function displayName(wrapper: string) {
  return (Component: React.ComponentType) =>
    `${wrapper}(${Component.displayName || Component.name || 'Component'})`;
}
