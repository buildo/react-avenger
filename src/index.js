import declareQueries from './queries';
import declareCommands from './commands';

export { declareQueries, declareCommands };
