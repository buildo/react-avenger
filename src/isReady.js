import every from 'lodash/every';
import map from 'lodash/map';

export default ({ readyState, ...props }) => {
  return every(map(readyState, (rs, k) => (
    props[k] !== void 0 && typeof rs.error === 'undefined'
  )));
};
