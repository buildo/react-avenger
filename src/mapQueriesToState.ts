import { RemoteData, RemotePending, RemoteSuccess, RemoteFailure } from './RemoteData';
import { Option, fromNullable } from 'fp-ts/lib/Option';
import { Queries } from '.';

type MappedState<Q extends Queries> = { [K in keyof Q]: RemoteData<string, Q[K]['_P']> };

function convertToRemoteData(
  data: Option<QuerySync>,
  prevValue: RemoteData<string, any>
): RemoteData<string, any> {
  return data.fold(prevValue, v => {
    if (v.loading) {
      if (prevValue._tag === 'RemoteSuccess') {
        return prevValue;
      }

      return new RemotePending();
    }

    if (v.error) {
      return new RemoteFailure(v.error);
    }

    return new RemoteSuccess(v.data);
  });
}

type QuerySync = { data?: any; loading: boolean; error: any };

export function mapQueriesToState<Q extends Queries>(
  { data }: { data: QuerySync },
  prevState: MappedState<Q>
): MappedState<Q> {
  return Object.entries(data).reduce((acc, [k, v]) => {
    return {
      ...acc,
      [k]: convertToRemoteData(fromNullable(v), prevState[k])
    };
  }, prevState);
}
