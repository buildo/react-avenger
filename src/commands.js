import React from 'react';
import debug from 'debug';
import pick from 'lodash/pick';
import mapValues from 'lodash/mapValues';
import t from 'tcomb';
import displayName from './displayName';

const warn = debug('react-avenger:commands');
warn.log = ::console.warn; // eslint-disable-line no-console

export const CommandsContextTypes = {
  commands: React.PropTypes.object
};

export default function commands(allCommands) {
  return (ids) => {
    if (process.env.NODE_ENV !== 'production') {
      ids.forEach(id => {
        if (!allCommands[id]) {
          warn(`command '${id}' not found! @commands declaration is: ${ids}`);
        }
      });
    }

    const CommandParamsTypes = ids.reduce((ac, k) => ({
      ...ac, ...(allCommands[k].params || {})
    }), {});

    const decorator = Component => {
      return class CommandsWrapper extends React.Component {
        static contextTypes = CommandsContextTypes;

        static displayName = displayName('commands')(Component);

        componentWillMount() {
          this._commands = mapValues(pick(this.context.commands, ids), cmd => params => {
            return cmd({ ...this.props, ...params });
          });
        }

        render() {
          return <Component {...this.props} {...this._commands}/>;
        }
      };
    };

    // we can't do much better here (i.e. no `maybe`)
    // commands can have actual params that we'll never be able to retrieve
    // implicitly / before component own lifecycle
    decorator.InputType = mapValues(CommandParamsTypes, ty => t.maybe(ty));
    decorator.OutputType = ids.reduce((ac, k) => ({ ...ac, [k]: t.Function }), {});
    decorator.Type = { ...decorator.InputType, ...decorator.OutputType };

    return decorator;
  };
}
