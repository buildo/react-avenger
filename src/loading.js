/* eslint-disable react/no-multi-comp */
import React from 'react';
import every from 'lodash/every';
import some from 'lodash/some';
import omit from 'lodash/omit';
import constant from 'lodash/constant';
import displayName from './displayName';

export const defaultIsLoading = ({ readyState }) => some(readyState, 'loading');

export const defaultIsReady = ({ readyState }) => every(readyState, 'ready');

export class NoWrapper extends React.Component { render = () => (this.props.children || [])[0] || null; }
export class NoLoader extends React.Component { render = () => null; }

export default function loading({
  isLoading = defaultIsLoading,
  isReady = defaultIsReady,
  wrapper = <NoWrapper />,
  loader = <NoLoader />,
  loaderProps = constant({}),
  wrapperProps = constant({})
} = {}) {

  return Component => class LoadingWrapper extends React.Component {

    static displayName = displayName('loading')(Component);

    render() {
      const ready = isReady(this.props);
      const loading = isLoading(this.props);
      const readyState = { ready, loading };
      return React.cloneElement(wrapper, wrapperProps(readyState), [
        ready && <Component {...omit(this.props, 'readyState')} key='content' />,
        loading && React.cloneElement(loader, {
          key: 'loader', ...loaderProps(readyState)
        })
      ]);
    }
  };
}

export const noLoaderLoading = loading();
