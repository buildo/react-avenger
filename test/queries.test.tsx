import * as React from 'react';
import { declareQueries } from '../src';
import { Query } from 'avenger';
import * as t from 'io-ts';
import { mount } from 'enzyme';

const makeQueries = () => {
  const foo = Query({
    params: { token: t.string },
    fetch: ({ token }) => Promise.resolve(token)
  });
  return { foo };
};

const sleep = (timeout: number) =>
  new Promise(resolve => {
    setTimeout(resolve, timeout);
  });

describe('declareQueries', () => {
  it('works', async () => {
    const { foo } = makeQueries();
    const render = jest.fn(JSON.stringify.bind(JSON));
    const WithFoo = declareQueries({ foo })(render as any);
    const mounted = mount(<WithFoo token="foo" />);
    expect(mounted).toMatchSnapshot();
    await sleep(10);
    expect(render.mock.calls.length).toBe(2);
    expect(render.mock.calls[0][0].foo._tag).toBe('RemotePending');
    expect(render.mock.calls[1][0].foo._tag).toBe('RemoteSuccess');
    expect(render.mock.calls[1][0].foo.value).toBe('foo');
  });

  it('works with querySync', async () => {
    const { foo } = makeQueries();
    const render = jest.fn(JSON.stringify.bind(JSON));
    const WithFoo = declareQueries({ foo }, { querySync: true })(render as any);
    mount(<WithFoo token="foo" />);
    await sleep(10);
    const mounted = mount(<WithFoo token="foo" />);
    expect(mounted).toMatchSnapshot();
    expect(render.mock.calls.length).toBe(3);
    expect(render.mock.calls[2][0].foo.value).toBe('foo');
  });
});
