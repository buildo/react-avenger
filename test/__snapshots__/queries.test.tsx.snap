// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`declareQueries works 1`] = `
<queries(mockConstructor)
  token="foo"
>
  <mockConstructor
    foo={
      RemotePending {
        "_tag": "RemotePending",
      }
    }
    token="foo"
  >
    {"token":"foo","foo":{"_tag":"RemotePending"}}
  </mockConstructor>
</queries(mockConstructor)>
`;

exports[`declareQueries works with querySync 1`] = `
<queries(mockConstructor)
  token="foo"
>
  <mockConstructor
    foo={
      RemoteSuccess {
        "_tag": "RemoteSuccess",
        "value": "foo",
      }
    }
    token="foo"
  >
    {"token":"foo","foo":{"value":"foo","_tag":"RemoteSuccess"}}
  </mockConstructor>
</queries(mockConstructor)>
`;
